src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,127 @@
+    private static final int INITIAL_ARRAY_SIZE = 10;
+    private static final double RESIZE_FACTOR = 1.5;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[INITIAL_ARRAY_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        expandCapacityIfArrayFull();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        expandCapacityIfArrayFull();
+        for (int i = size - 1; i >= index; i--) {
+            elementData[i + 1] = elementData[i];
+        }
+        elementData[index] = value;
+        size++;
+
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("There is no index: " + index);
+        }
+    }
 
+    private void checkIndexForGet(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("There is no index: " + index);
+        }
+    }
+
+    private void expandCapacityIfArrayFull() {
+        if (size == elementData.length) {
+            int newCapacity = (int) (elementData.length * RESIZE_FACTOR);
+            Object[] newArray = new Object[newCapacity];
+            System.arraycopy(elementData, 0, newArray, 0, size);
+            elementData = (T[]) newArray;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); ++i) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForGet(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndexForSet(index);
+        elementData[index] = value;
+    }
 
+    private void checkIndexForSet(int index) {
+        if (index >= 0 && index <= size - 1) {
+            return;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("There is no index: " + index);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForRemove(index);
+        T removedValue = elementData[index];
+        for (int i = index; i < size - 1; i++) {
+            elementData[i] = elementData[i + 1];
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = indexOf(element);
+        if (index != -1) {
+            return remove(index);
+        } else {
+            throw new NoSuchElementException("There's no element: " + element);
+        }
+    }
+
+    private void checkIndexForRemove(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("There is no index: " + index);
+        }
+    }
```

Comment:
no need to duplicate methods logic just because of naming problem, better choose another name
```suggestion
```